### PR TITLE
rate-limiting policies description and usage

### PR DESCRIPTION
**Pull request by Tieske**

### CODE REVIEW

File: app/plugins/rate-limiting.md [hunk 1]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
 {"message":"API rate limit exceeded"}
 ```
 
+## Implementation considerations
+
+The plugin supports 3 policies, which each have their specific pro's and con's.
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: >which each have their specific `pros` and `cons`.

File: app/plugins/rate-limiting.md [hunk 2]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
...
+
+policy    | pro's          | con's
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: pros | cons

File: app/plugins/rate-limiting.md [hunk 3]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
...
+---       | ---            | ---
+`cluster` | accurate, no extra components to support  | relatively the biggest performance impact, each request forces a read and a write on the underlying datastore.
+`redis`   | accurate, lesser performance impact than a `cluster` policy | extra redis installation to support, bigger performance impact than a `local` policy
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: > extra redis installation `required`, bigger performance impact...

File: app/plugins/rate-limiting.md [hunk 4]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
...
+
+**NOTE**: the redis policy does not support the Sentinel protocol for high available
+master-slave architectures. When using rate-limiting for general protection the chances
+of both redis being down and the system being under attack are rather small. Check
+with your own use case wether you can handle this (small) risk.
+
+### Every transaction counts
+
+In this scenario, the `local` policy is not an option. So here the decision is between
+the extra performance of the `redis` policy against its extra support effort. Based on that balance
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: >Based on that `balance,` ...

File: app/plugins/rate-limiting.md [hunk 5]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
...
+the choice should either be `cluster` or `redis`.
+
+The recommendation is to start with the `cluster` policy, with the option to move over to `redis`
+if performance takes too big a hit. The thing to keep in mind is that existing usage metrics cannot
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: >if performance `reduces drastically`. `Keep in mind existing usage` metrics cannot ...

File: app/plugins/rate-limiting.md [hunk 6]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
...
+be ported from the datastore to redis. Generally with shortlived metrics (per second or per minute)
+this is not an issue, but with longer lived ones (months) it might be, so you might want to plan
+your switch more carefully.
+
+### Backend protection
+
+As accuracy is of lesser importance, the `local` policy can be used. It might require some experimenting
+to get the proper setting. For example if the user is bound to 100 requests per second, and you have a
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: > For `example,` ... and you have `an` ...

File: app/plugins/rate-limiting.md [hunk 7]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
...
+equally balanced 5 node kong Cluster, setting the `local` limit to something like 30 requests per second
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: > ... 5 node `Kong` Cluster ...

File: app/plugins/rate-limiting.md [hunk 8]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
...
+should work. If you are worried about too many false-negatives, increase the value.
+
+Most likely the user will be granted more than was agreed, but it will effectively block any attacks, while
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: > ...block any `attacks while` ...

File: app/plugins/rate-limiting.md [hunk 9]
```diff
@@ -78,6 +78,53 @@ If any of the limits configured is being reached, the plugin will return a `HTTP
...
+maintaining the best performance. The thing to keep in mind here is that when the cluster scales to more
```
Comment (theetrain) [inline app/plugins/rate-limiting.md]: > ... performance. `Keep in mind as the cluster scales` to more `nodes,` the users will ...

**Comment by Tieske** [comment]
add a note that using a consistent hashing LB in front of Kong would significantly increase consistency of the `local` policy

**Pull request merged**